File: python/leetcode/200-number-of-islands.py
class Solution:
    def dfs(self, grid, r, c):
        rows, cols = len(grid), len(grid[0])

        if (r < 0 or r >= rows or c < 0 or c >= cols or grid[r][c] == '0'):
            return
        grid[r][c] = '0'
        self.dfs(grid, r+1, c)
        self.dfs(grid, r-1, c)
        self.dfs(grid, r, c+1)
        self.dfs(grid, r, c-1)

    def numIslands(self, grid: List[List[str]]) -> int:
        num_islands = 0

        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col] == '1':
                    num_islands += 1
                    self.dfs(grid, row, col)

        return num_islands


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        num_islands = 0
        ROWS, COLS = len(grid), len(grid[0])
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        def dfs(r, c):
            if r < 0 or r >= ROWS or c < 0 or c >= COLS or grid[r][c] != "1":
                return
            grid[r][c] = "0"
            for dr, dc in directions:
                nr, nc = r+dr, c+dc
                dfs(nr, nc)

        for row in range(ROWS):
            for col in range(COLS):
                if grid[row][col] == "1":
                    num_islands += 1
                    dfs(row, col)
        return num_islands

File: c#/leetcode/200-number-of-islands.cs
public class Solution
{
  public int NumIslands(char[][] grid)
  {
    int numIslands = 0;
    for (int i = 0; i < grid.Length; i++)
    {
      for (int j = 0; j < grid[0].Length; j++)
      {
        if (grid[i][j] == '1')
        {
          numIslands++;
          dfs(grid, i, j);
        }
      }
    }
    return numIslands;
  }

  public void dfs(char[][] grid, int r, int c)
  {ds
    int rows = grid.Length, cols = grid[0].Length;

    if (r < 0 || r >= rows || c < 0 || c >= cols || grid[r][c] == '0')
    {
      return;
    }

    grid[r][c] = '0';
    dfs(grid, r + 1, c);
    dfs(grid, r - 1, c);
    dfs(grid, r, c + 1);
    dfs(grid, r, c - 1);
  }
}